torch/futures/__init__.py

```diff
@@ -81,6 +81,39 @@ def then(self, callback):  # type: (Callable[[Future[T]], S]) -> Future[S]
         """
         return cast(Future[S], super().then(callback))
 
+    # Have to use string annotations because  PEP-0563 is not available in 3.6
+    def add_done_callback(self, callback):  # type: (Callable[[Future[T]], None]) -> None
+        r"""
+        Add a callback to this ``Future``.
+        Multiple callbacks can be added to the same ``Future``, and all
+        added callbacks will be executed once the future completes.
+        If the future has already completed, the callback will be executed immediately.
+
+        Append the given callback function to this ``Future``, which will be run
+        when the ``Future`` is completed.  Multiple callbacks can be added to
+        the same ``Future``, and will be invoked in the same order as they were
+        added. The callback must take one argument, which is the reference to
```

Comment:
Shall we add a test where the callback function does not take any argument and verifies that the error message is clear?

Reply:
Done. The callback logs exceptions to stdout, and I had trouble redirecting the content of stdout that was emitted from C++ in the python test so I could assert it's contents- right now the test just asserts that no exception was thrown. Here's an example though (showing that the exception is logged to stdout, but no exception is raised in the main thread)

![image](https://user-images.githubusercontent.com/16311747/95229334-730fba00-07ce-11eb-8b82-f7d099749a31.png)

Reply:
Can the following code solve this problem?

https://github.com/pytorch/pytorch/blob/275bb5e80139efc6f779b09d342b39eacccd97a3/torch/csrc/jit/python/pybind_utils.h#L127-L142

Looks like it works for existing tests:

https://github.com/pytorch/pytorch/blob/275bb5e80139efc6f779b09d342b39eacccd97a3/test/test_futures.py#L111-L116

Reply:
I think the difference there is that when a callback registered with `then` encounters an exception, that exception gets bubbled up to the thread calling `then_fut.wait()` (making it easy to assert in the test).

Meanwhile the docs for `add_done_callback` assert that any exceptions encountered during the execution of the callback should be logged and ignored (https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.Future.add_done_callback)- so we don't want to assert that we encountered an exception, we'd want to assert some string getting sent to stdout. I tried out this: https://stackoverflow.com/questions/4219717/how-to-assert-output-with-nosetest-unittest-in-python, but from some testing it only looks like it doesn't pick up anything that I log from C++ code (which is where I'm logging the exception).

Reply:
I think the difference there is that when a callback registered with `then` encounters an exception, that exception gets bubbled up to the thread calling `then_fut.wait()` (making it easy to assert in the test).

Meanwhile the docs for `add_done_callback` assert that any exceptions encountered during the execution of the callback should be logged and ignored (https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.Future.add_done_callback)- so we don't want to assert that we encountered an exception, we'd want to assert some string getting sent to stdout. I tried out this: https://stackoverflow.com/questions/4219717/how-to-assert-output-with-nosetest-unittest-in-python, but from some testing it only looks like it doesn't pick up anything that I log from C++ code (which is where I'm logging the exception).